### PR TITLE
fix: constant fold in `SubstituteExp`

### PR DESCRIPTION
**Pull request by plun411**
completes task 1 in #3638

### CODE REVIEW

**Comment by magnus-madsen** [comment]
@mlutze Why does the greeting fail?

**Comment by mlutze** [comment]
Ugh not sure. There's a big ticket about this error since everyone and their dog is getting it.

Not sure this bot is worth the trouble

File: main/src/ca/uwaterloo/flix/language/phase/Inliner.scala [hunk 1]
```diff
@@ -209,22 +209,14 @@ object Inliner {
...
       val e1 = visitExp(exp1, subst0)
       val e2 = visitExp(exp2, subst0)
-      (sop, e1, e2) match {
-        case (SemanticOperator.BoolOp.And, LiftedAst.Expression.False(_), _) => LiftedAst.Expression.False(loc)
-        case (SemanticOperator.BoolOp.And, _, LiftedAst.Expression.False(_)) if e1.purity == Pure => LiftedAst.Expression.False(loc)
-        case (SemanticOperator.BoolOp.Or, LiftedAst.Expression.True(_), _) => LiftedAst.Expression.True(loc)
-        case (SemanticOperator.BoolOp.Or, _, LiftedAst.Expression.True(_)) if e1.purity == Pure => LiftedAst.Expression.True(loc)
-        case _ => LiftedAst.Expression.Binary(sop, op, e1, e2, tpe, purity, loc)
-      }
+      val binExp = LiftedAst.Expression.Binary(sop, op, e1, e2, tpe, purity, loc)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Inliner.scala:218]: For both  cases, you should *not* construct a new (possibly temporary) exp just to be pass it as an argument.

In FP in general that is considered an anti pattern. One reason is that it makes it harder to understand termination. 

Instead, you should pass in each component, e,g here would be `sop`, `op`, `e1`, `e2`, and so forth, and then let binaryFold do its job.

File: main/src/ca/uwaterloo/flix/language/phase/Inliner.scala [hunk 2]
```diff
@@ -501,6 +493,42 @@ object Inliner {
...
+      case (SemanticOperator.BoolOp.Or, _, LiftedAst.Expression.True(_)) if exp0.exp1.purity == Pure => LiftedAst.Expression.True(exp0.loc)
+      case _ => exp0
+    }
+  }
+
+  /**
+   * Performs boolean folding on a given unary expression with the logic:
+   * Folds not-true => false and not-false => true.
+   */
+  private def unaryFold(exp0: LiftedAst.Expression.Unary): LiftedAst.Expression = {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Inliner.scala:524]: This should occur before binaryFold to keep the right order.

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/Inliner.scala:524]: I think both of these helper functions should occur further down in the file, towards the bottom.

File: main/src/ca/uwaterloo/flix/language/phase/Inliner.scala (hunk 1 above)
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/Inliner.scala:218]: Ask me next time, and I will try to better explain this point :)

File: main/src/ca/uwaterloo/flix/language/phase/Inliner.scala [hunk 3]
```diff
@@ -501,6 +493,42 @@ object Inliner {
...
+   * Performs boolean folding on a given binary expression with the logic:
+   * Only fold if the expression removed is pure, and
+   * For Or-expressions,
+   * fold into true if either the left or right expression is true.
+   * if either the left or right expression is false, fold into the other expression.
+   * For And-expressions,
+   * fold into false, if either the left or right expression is false.
+   * if either the left or right expression is true, fold into the other expression.
+   */
+  private def binaryFold(exp0: LiftedAst.Expression.Binary): LiftedAst.Expression = {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Inliner.scala:506]: This function should *not* take a binary expression (as explained above), but rather take these arguments:

sop, op, e1, e2, tpe, purity, loc

**Review by magnus-madsen — COMMENTED** [review]
Added some comments.

**Comment by magnus-madsen** [comment]
> Ugh not sure. There's a big ticket about this error since everyone and their dog is getting it.
> 
> Not sure this bot is worth the trouble

Maybe its time to the let bot rest in pieces?

**Comment by magnus-madsen** [comment]
Thanks!

**Pull request merged**